### PR TITLE
Require `Debug` for `AnalyzerRule`, `FunctionRewriter`, and `OptimizerRule`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/12555

## Rationale for this change

I would like DataFusion to be easiser to configure and use, and having `Debug` impls for
common structs would be a big help.

## What changes are included in this PR?

1. Require `Debug` for `AnalyzerRule`, `FunctionRewriter`, and `OptimizerRule`
2. `#derive` this trait as necessary

## Are these changes tested?
Yes buy CI and the compiler

## Are there any user-facing changes?
Technically this is an API change as it also adds the requirement for Debug for various optimizer rules, but I think that is  good thing.

### CODE REVIEW

File: datafusion-examples/examples/optimizer_rule.rs [hunk 1]
```diff
@@ -100,6 +100,7 @@ pub async fn main() -> Result<()> {
 
 /// An example OptimizerRule that replaces all `col = <const>` predicates with a
 /// user defined function
+#[derive(Default, Debug)]
```
Comment (alamb) [inline datafusion-examples/examples/optimizer_rule.rs:103]: this is the end user visible change -- basically you have to `#[derive(Debug)]` if you haven't already

File: datafusion/optimizer/src/analyzer/mod.rs [hunk 2]
```diff
@@ -72,7 +74,7 @@ pub trait AnalyzerRule {
 ///
 /// An `Analyzer` transforms a `LogicalPlan`
 /// prior to the rest of the DataFusion optimization process.
-#[derive(Clone)]
+#[derive(Clone, Debug)]
```
Comment (alamb) [inline datafusion/optimizer/src/analyzer/mod.rs:77]: It will be nice to be able to see what rules are registered with the analyzer/optimizer

**Review by phillipleblanc — APPROVED** [review]

**Review by goldmedal — APPROVED** [review]
Thanks @alamb It's a nice change for the user who wants to implement the custom rules 👍

**Review by jayzhan211 — APPROVED** [review]
👍

**Comment by alamb** [comment]
🚀

**Pull request merged**